File: Shared/Resources/radixsort.bundle/radixsort.rb
def counting(arr, e)
  n = arr.length()
  output = Array.new(n) { |z| 0 }
  count = Array.new(10) { |z| 0 }
  for i in (0...n)
    idx = arr[i] / e
    count[idx % 10] += 1
  end
  for i in (1...10)
    count[i] += count[i - 1]
  end
  i = n - 1
  while i >= 0
    idx = arr[i] / e
    output[count[idx % 10] - 1] = arr[i]
    count[idx % 10] -= 1
    i -= 1
  end
  i = 0
  for i in (0...n)
    arr[i] = output[i]
  end
end

def sort(arr)
  x = arr.max()
  e = 1
  while x / e > 0
    counting(arr, e)
    e *= 10
  end
end

array = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(array)
p array

File: App/Resources/AlgorithmDetails/radixsort.bundle/radixsort.c
#include <stdio.h>
#include <stdlib.h>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23,
                 90, 69, 51, 81, 68, 83, 32, 56};

void printList(int items[], int size) {
  for (int i = 0; i < size; i++) {
    if (i == 0) {
      printf("[%d, ", items[i]);
    } else if (i != size - 1) {
      printf("%d, ", items[i]);
    } else {
      printf("%d]", items[i]);
    }
  }
}

int getMax(int arr[], int n) {
  int max = arr[0];
  for (int i = 1; i < n; i++) {
    if (arr[i] > max) {
      max = arr[i];
    }
  }
  return max;
}

void sort(int arr[], int n) {
  int bucket[10][10];
  int bucketCount[10];
  int i, j, k, r, nop = 0, divisor = 1, lar, pass;
  lar = getMax(arr, n);
  while (lar > 0) {
    nop++;
    lar /= 10;
  }
  for (pass = 0; pass < nop; pass++) {
    for (i = 0; i < 10; i++) {
      bucketCount[i] = 0;
    }
    for (i = 0; i < n; i++) {
      r = (arr[i] / divisor) % 10;
      bucket[r][bucketCount[r]] = arr[i];
      bucketCount[r] += 1;
    }
    i = 0;
    for (k = 0; k < 10; k++) {
      for (j = 0; j < bucketCount[k]; j++) {
        arr[i] = bucket[k][j];
        i++;
      }
    }
    divisor *= 10;
  }
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}